base_layer/core/src/proof_of_work/monero_merkle_tree/monero_merkle_tree.rs

```diff
@@ -0,0 +1,248 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::proof_of_work::monero_merkle_tree::{
+    monero_merkle_element::MoneroMerkleElement,
+    monero_merkle_hash_util::{create_leaf_hash, create_node_hash},
+    monero_merkle_proof::MoneroMerkleProof,
+};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::{vec_deque::Iter, BTreeMap, VecDeque},
+    rc::Rc,
+};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum MoneroMerkleProofNode {
+    Left(Vec<u8>),
+    Right(Vec<u8>),
+}
+
+/// MerkleTree struct represents merkle binary tree with values of type Vec<u8> and map of nodes.
+#[derive(Debug, Clone)]
+pub struct MoneroMerkleTree {
+    root: MoneroMerkleElement,
+    height: usize,
+    count: usize,
+    storage: VecDeque<Rc<Vec<u8>>>,
+    nodes: BTreeMap<usize, VecDeque<MoneroMerkleElement>>,
+}
+
+impl MoneroMerkleTree {
+    /// Creates new, empty `MerkleTree`.
+    pub fn new() -> Self {
+        MoneroMerkleTree {
+            root: MoneroMerkleElement::empty(),
+            height: 0,
+            count: 0,
+            storage: VecDeque::new(),
+            nodes: BTreeMap::new(),
+        }
+    }
+
+    /// Creates `MerkleTree` from `Vec` of u8.
+    pub fn from_vec(data: Vec<Vec<u8>>) -> Self {
```

Comment:
The MerkleTree class was removed in #2061 since the is no way to request a proof directly from MoneroD. Ported the tree_hash algorithm from MoneroD in its' place.